File: csi/java/src/main/java/org/shypl/biser/csi/ConnectionCloseReason.java
package org.shypl.biser.csi;

public enum ConnectionCloseReason {
	NONE,
	SERVER_SHUTDOWN,
	ACTIVITY_TIMEOUT_EXPIRED,
	AUTHORIZATION_REJECT,
	RECOVERY_REJECT,
	CONCURRENT,
	PROTOCOL_BROKEN,
	SERVER_ERROR;

	public static byte getProtocolFlag(ConnectionCloseReason reason) {
		switch (reason) {
			case NONE:
				return Protocol.CLOSE;
			case SERVER_SHUTDOWN:
				return Protocol.CLOSE_SERVER_SHUTDOWN;
			case ACTIVITY_TIMEOUT_EXPIRED:
				return Protocol.CLOSE_ACTIVITY_TIMEOUT_EXPIRED;
			case AUTHORIZATION_REJECT:
				return Protocol.CLOSE_AUTHORIZATION_REJECT;
			case RECOVERY_REJECT:
				return Protocol.CLOSE_RECOVERY_REJECT;
			case CONCURRENT:
				return Protocol.CLOSE_CONCURRENT;
			case PROTOCOL_BROKEN:
				return Protocol.CLOSE_PROTOCOL_BROKEN;
			case SERVER_ERROR:
				return Protocol.CLOSE_SERVER_ERROR;
			default:
				throw new IllegalArgumentException();
		}
	}
}

File: kotlin/biser-kotlin-js/src/main/kotlin/org/shypl/biser/csi/ConnectionCloseReason.kt
package org.shypl.biser.csi

enum class ConnectionCloseReason {
	NONE,
	SERVER_SHUTDOWN,
	ACTIVITY_TIMEOUT_EXPIRED,
	AUTHORIZATION_REJECT,
	RECOVERY_REJECT,
	RECOVERY_TIMEOUT_EXPIRED,
	CONCURRENT,
	PROTOCOL_BROKEN,
	SERVER_ERROR;
	
	companion object {
		fun getReason(protocolFlag: Byte): ConnectionCloseReason? {
			return when (protocolFlag) {
				Protocol.CLOSE                          -> NONE;
				Protocol.CLOSE_SERVER_SHUTDOWN          -> SERVER_SHUTDOWN;
				Protocol.CLOSE_ACTIVITY_TIMEOUT_EXPIRED -> ACTIVITY_TIMEOUT_EXPIRED;
				Protocol.CLOSE_AUTHORIZATION_REJECT     -> AUTHORIZATION_REJECT;
				Protocol.CLOSE_RECOVERY_REJECT          -> RECOVERY_REJECT;
				Protocol.CLOSE_CONCURRENT               -> CONCURRENT;
				Protocol.CLOSE_PROTOCOL_BROKEN          -> PROTOCOL_BROKEN;
				Protocol.CLOSE_SERVER_ERROR             -> SERVER_ERROR;
				else                                    -> null
			}
		}
	}
}